[aten] Split some at::launch code into at::internal::launch_no_thread_state()

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38477 [aten] Split some at::launch code into at::internal::launch_no_thread_state()**

A few specific uses (e.g. Thrift rpc parsing) don't need source thread
state to be copied over. In microbenchmarks, this seems to add ~500ns,
so split code across functions, so some code can use directly.

Differential Revision: [D21573168](https://our.internmc.facebook.com/intern/diff/D21573168/)

Commit:
[aten] Split some at::launch code into at::internal::launch_no_thread_state()

A few specific uses (e.g. Thrift rpc parsing) don't need source thread
state to be copied over. In microbenchmarks, this seems to add ~500ns,
so split code across functions, so some code can use directly.

Differential Revision: [D21573168](https://our.internmc.facebook.com/intern/diff/D21573168/)

[ghstack-poisoned]

jjlilley commented:
fwiw: the motivation for this change is simply to improve efficiency of D21409536

wanchaol commented on aten/src/ATen/Parallel.h:
```diff
@@ -128,6 +128,9 @@ CAFFE2_API int get_num_interop_threads();
 // Launches inter-op parallel task
 CAFFE2_API void launch(std::function<void()> func);
+namespace internal {
+  void launch_no_thread_state(std::function<void()> fn);
```
nit: align the `func` name with `launch`

jjlilley replied:
sorry, clash of different style guidelines in my head :) fixed.

Commit:
fix indentation on "[aten] Split some at::launch code into at::internal::launch_no_thread_state()"


A few specific uses (e.g. Thrift rpc parsing) don't need source thread
state to be copied over. In microbenchmarks, this seems to add ~500ns,
so split code across functions, so some code can use directly.

Differential Revision: [D21573168](https://our.internmc.facebook.com/intern/diff/D21573168/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good to me.
